### ISSUE TITLE
Show RHEL in the machines list even when downloads is selected?

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit-machines#269]
Right now, it's not obvious why RHEL is not available in the OS list.

We probably still should show RHEL in the list, even though it's not downloadable.  

Even when the option is selected, we could have a little notice saying it's not automatically downloadable and another installation method should be selected to use RHEL in Cockpit-Machines, along with a link on where to download RHEL (or how to make it work with Satellite, which I think would use a local URL, for example).

Perhaps we should show the same thing for other OSes that are not directly downloadable too. (Having some sort of way to show that an OS is not downloadable in the list itself would be ideal. We may need to use a non-form PF4 select for this capability, otherwise we could append some text like "(not downloadable)").

---

### Discussed in https://github.com/cockpit-project/cockpit-machines/discussions/268

<div type='discussions-op-text'>

<sup>Originally posted by **Verhoeckx** July 22, 2021</sup>
Hello Cockpit developers,

I see that Red Hat Enterprise Linux 8.x is not in the list of installable operating systems in the window Create new Virtual Machine?
Why not? Shouldn't it be there? It's the main operating system of Red Hat!

I guess it's not a problem for me because I already downloaded Red Hat Enterprise Linux 8.4 on my local machine. I just have to copy it to the server.</div>

**Comment by KKoukiou** [comment]
The issue RHEL is not appearing in the downloads list is that it's missing the installation tree URL in the osinfo DB.

See fedora: https://gitlab.com/libosinfo/osinfo-db/-/blob/master/data/os/fedoraproject.org/fedora-33.xml.in#L161

@fidencio is this because we can't just have the URLs open upstream? Is there another package for RHEL users maybe? I remember we had a discussion when we implemented this, sorry for repeating.

**Comment by fidencio** [comment]
> The issue RHEL is not appearing in the downloads list is that it's missing the installation tree URL in the osinfo DB.
> 
> See fedora: https://gitlab.com/libosinfo/osinfo-db/-/blob/master/data/os/fedoraproject.org/fedora-33.xml.in#L161
> vs: https://gitlab.com/libosinfo/osinfo-db/-/blob/master/data/os/redhat.com/rhel-8.4.xml.in#L49
> 
> @fidencio is this because we can't just have the URLs open upstream?

That's exactly the case.  Red Hat doesn't provide any direct download link for the medias, nor the trees, that osinfo could use.

> Is there another package for RHEL users maybe?

No, unfortunately not.  There's a package for developers **inside** Red Hat, and that points to our internal mirrors, but that doesn't o to customers at all, as it relies on access to our VPN.

> I remember we had a discussion when we implemented this, sorry for repeating.

No problem, it's good that you brought this up and we can have this registered for posterity here. :-)

**Comment by KKoukiou** [comment]
Thanks @fidencio!

@garrett that said RHEL will  not be available for download. I don't think however we should handle it differently than other distros and show a info/warning that it's not available for auto-download. We could maybe consider adding a note in the dialog, mentioning that one can only see the OSes which have a installation media tree URL in osinfo db. Although that's already a technical detail.

Please consider the option and let me know, if we can either close this issue or add some kind of generic (for all missing distros) note.

**Comment by garrett** [comment]
So, there are two things:

1. People expect Red Hat Enterprise Linux (RHEL) in Cockpit, especially when using Cockpit on RHEL itself.
2. There isn't an easy way to download RHEL, without a subscription or an account. But there are links to download RHEL with an account (even a free account).

Since the Cockpit team is employed by Red Hat and we ship Cockpit as Web Console as a part of RHEL, it's probably fine (and even a good idea) to special case this one exception. All the other Linux distros are free to download — with a few exceptions, like:
- Elementary, where it costs to download, but that can manually be set to $0... but it still doesn't require a log in.
- SUSE Enterprise Linux — which, like RHEL, requires signing in to download

The most streamlined way to handle this in an expected manner is to probably have it in the dropdown (possibly with some sort of signifier to show it's different?) and when "Red Hat Enterprise..." is selected, have a text block below that explains it's not possible to download directly, with a link to one of the following places to download it:

- https://developers.redhat.com/products/rhel/download
- https://access.redhat.com/downloads/

And then a suggestion about how to use the downloaded image to create a VM in Cockpit.

I say we strongly consider adding this one-off for RHEL specifically. (And if SUSE, the only other Linux distro that I'm aware of that requires an account to download, wants to follow along with a one-off, they can send a patch to do something similar. But if we have more than just RHEL being a placeholder with a message in the downloader area, perhaps we might want to show that it cannot be directly downloaded somehow in the dropdown itself.)

And, at some point in the future, we should have a streamlined method to actually download RHEL directly when Cockpit is running on a registered RHEL system.

**Comment by fidencio** [comment]
@garrett, let me loop in @felipeborges from GNOME Boxes, as he did an amazing work on integrating what seems to be either something really close to your suggestion (if not exactly your suggestion) on their software.  And I think a lot can be shared from the work he did there.

**Comment by felipeborges** [comment]
GNOME Boxes downloads RHEL from the developers.redhat.com portal pointed aboved by Garret.

TL;DR: We popup a webview for the user to enter credentials in the portal's "download manager" application, which gets us a final download URL tied to the user session.

We use https://developers.redhat.com/download-manager/rest/featured/file/rhel?tag=$OUR_CUSTOM_USER_AGENT_FOR_METRICS as an authentication URI. After the user authenticates properly in it, the page will redirect to the latest RHEL media ISO (some address in access.cdn.redhat.com/....).

What Boxes does is that it connects to the `uri` property of the webview and when it matches the cnd prefix we are able to query the `tcDownloadURL` parameter from the request, which contains the final URL we need.

See https://gitlab.gnome.org/GNOME/gnome-boxes/-/blob/master/src/assistant/rhel-download-dialog.vala#L39 where we load the authentication page, and https://gitlab.gnome.org/GNOME/gnome-boxes/-/blob/master/src/assistant/rhel-download-dialog.vala#L55 where we filter the redirection request until it satisfies our need for a download URL.

**Comment by KKoukiou** [comment]
This is now supported

![Screen Shot 2023-01-05 at 20 39 50](https://user-images.githubusercontent.com/14921356/210865928-2b2db892-0a11-44ee-946c-b2c7f933b802.png)
